Snabbwall upstreaming PR

takikawa opened the pull request:
Hi Kat, here's the PR for upstreaming Snabbwall.

Most of the changes are in `src/apps/wall` and `src/program/wall`. There are selftests for some things in apps, and also more end-to-end style tests in `src/program/wall/tests`.

This also makes some changes in pflua that were needed in order to write firewall rules using `pfmatch`. If you have any feedback on that (like should it also be separately merged in `Igalia/pflua` as well? Or does the change look ok?) that would be great.

The other main addition is the `ljndpi` library in `lib` (written by @aperezdc, I don't think I changed anything in there myself in the milestones I worked on).

aperezdc committed:
Add a selftest() function to apps.wall.scanner

aperezdc committed:
Discard Ethernet header when scanning IPv6 headers

This allows passing an offset to ipv6_walk_headers(), which allows to take
into account that packets have Ethernet headers at the start (possibly with
VLAN tags).

aperezdc committed:
Add comment for the ihl() function

Commit:
Fix GCC 6 -Wmisleading-indentation warnings.

Thanks to Roman Tsisyk.

Commit:
Merge branch 'master' into v2.1

Commit:
Constrain value range of lj_ir_kptr() to unsigned 32 bit pointers.

Thanks to Peter Cawley.

aperezdc committed:
Merge pull request #8 from aperezdc/swall-starfruit-wip

Implement the L7Spy application and "wall spy" command

aperezdc committed:
Remove unneeded assertions on flow key struct sizes

The assertion is done in make_cdata_hash_function(), so it is not needed to do
it before calling it. This removes a couple of local variables as well.

aperezdc committed:
Combine intermediate hash result while iterating over the input

This makes the hash functions generated by make_cdata_hash_function()
use the intermediate value calculated for the previous 32-bit chunks
of the input combined with the output for the current chunk. Without
this fix, Changes at the beginning of the input would not make the
final hash result vary, which was incorrect behavior.

aperezdc committed:
Do not clear the last bit of the calculated hash value

Contrary to the version in lib.ctable, the hash32() function here does not
need to clear the last bit of the result, because a "guard" value is unneeded
for flow keys.

Commit:
Use the GDB JIT API in a thread-safe manner.

Thanks to Peter Cawley.

Commit:
x86/x64: Fix instruction length decoder.

Thanks to Peter Cawley.

Commit:
Merge branch 'master' into v2.1

Commit:
Windows/x86: Add full exception interoperability.

Contributed by Peter Cawley.

Commit:
Fix dependencies.

Commit:
Remove assumption that lj_math_random_step() doesn't clobber FPRs.

Commit:
Merge branch 'master' into v2.1

Commit:
Workaround for MinGW headers lacking some exception definitions.

Commit:
Add guard for obscure aliasing between open upvalues and SSA slots.

Thanks to Peter Cawley.

Commit:
Merge branch 'master' into v2.1

Commit:
LJ_GC64: Introduce IRT_PGC.

Contributed by Peter Cawley.

Commit:
Move common 32/64 bit in-memory FP constants to jit_State.

Prerequisite for immovable IR.
Contributed by Peter Cawley.

Commit:
Add IR_FLOAD with REF_NIL for field loads from GG_State.

Contributed by Peter Cawley.

Commit:
Load SIMD constants with IR_FLOAD from GG_State.

Contributed by Peter Cawley.

Commit:
Add ra_addrename().

Contributed by Peter Cawley.

Commit:
Make the IR immovable after assembly.

This allows embedding pointers to IR constants in the machine code.
Contributed by Peter Cawley.

Commit:
Simplify GCtrace * reference embedding for trace stitching.

This is now possible due to the immovable IR.
Contributed by Peter Cawley.

Commit:
Always walk IR constants in ascending order.

Prerequisite for embedding 64 bit constants directly in the IR.
Contributed by Peter Cawley.

Commit:
Embed 64 bit constants directly in the IR, using two slots.

Contributed by Peter Cawley.

Commit:
Strip out old infrastructure for 64 bit constants.

Contributed by Peter Cawley.

Commit:
LJ_GC64: Add support for 64 bit GCobj constants in the IR.

Contributed by Peter Cawley.

Commit:
LJ_GC64: Update IR type sizes.

Contributed by Peter Cawley.

Commit:
LJ_FR2: Add support for trace recording and snapshots.

Contributed by Peter Cawley.

Commit:
x64/LJ_GC64: Add missing backend support and enable JIT compilation.

Contributed by Peter Cawley.

Commit:
x86: Don't spill an explicit REF_BASE in the IR.

Thanks to Vyacheslav Egorov.

Commit:
Merge branch 'master' into v2.1

Commit:
Use MAP_TRYFIXED for the probing memory allocator, if available.

Commit:
Cleanup install docs.

Commit:
Fix collateral damage from LJ_GC64 changes to asm_href().

Commit:
x64/LJ_GC64: Fix __call metamethod for tailcall.

Commit:
Fix compiler warning.

Commit:
DynASM/MIPS: Add missing MIPS64 instructions.

Contributed by Djordje Kovacevic and Stefan Pejic from RT-RK.com.
Sponsored by Cisco Systems, Inc.

Commit:
MIPS64, part 1: Add MIPS64 support to interpreter.

Contributed by Djordje Kovacevic and Stefan Pejic from RT-RK.com.
Sponsored by Cisco Systems, Inc.

Commit:
Proper fix for LJ_GC64 changes to asm_href().

Commit:
MIPS: Fix build failures and warnings.

Commit:
Don't try to record outermost pcall() return to lower frame.

Commit:
Merge branch 'master' into v2.1

Commit:
Fix Valgrind suppressions.

Commit:
Merge branch 'master' into v2.1

Commit:
LJ_GC64: Allow optional use of the system memory allocator.

Commit:
LJ_GC64: Ensure all IR slot fields are initialized.

Commit:
LJ_GC64: Set correct nil value when clearing a cdata finalizer.

Thanks to Stefan Pejic.

Commit:
Fix PHI remarking in SINK pass.

Thanks to Vyacheslav Egorov.

Commit:
Merge branch 'master' into v2.1

Commit:
x64/LJ_GC64: Fix code generation for IR_KNULL call argument.

Commit:
MIPS: Support MIPS16 interlinking.

Commit:
Drop leftover regs in 'for' iterator assignment, too.

Commit:
Merge branch 'master' into v2.1

Commit:
Properly clean up state before restart of trace assembly.

Commit:
Fix unused vars etc. in internal Lua files.

Thanks to François Perrad.

Commit:
Merge branch 'master' into v2.1

Commit:
Fix for cdata vs. non-cdata arithmetics/comparisons.

Thanks to Vyacheslav Egorov.

Commit:
Merge branch 'master' into v2.1

Commit:
Adjust comment with defines.

Commit:
Set arg table before evaluating LUA_INIT and -e chunks.

Commit:
Emit bytecode in .c/.h files with unsigned char type.

Commit:
Merge branch 'master' into v2.1

aperezdc committed:
Merge tag 'v2016.07' into snabbwall

aperezdc committed:
Add a __tostring metamethod for protocol_bitmask values

This allows for having a human-readable representation, which can come
in handy, mostly while debugging.

aperezdc committed:
Change log: Fix formatting.

aperezdc committed:
Expose the nDPI version and use versioned ndpi.protocol_ids

This obtains at runtime the version of the nDPI library using the
ndpi_revision() function, and exposes it as a table at "ndpi.lib_version".
This table has fields .major, .minor, and .patch with their usual meanings.

Additionally, the existing ndpi.protocol_ids module is renamed to
ndpi.protocol_ids_1_7, and a new ndpi.protocol_ids_1_8 module is added
(created using tools/update-protocol-ids). The detected library version
numbers are used to determine which one of the versioned protocol_ids
submodule has to be loaded.

This patch is a first step towards supporting nDPI 1.8 (issue #1).

Commit:
Must preserve J->fold.ins (fins) around call to lj_ir_ksimd().

Commit:
Fix exit status for 'luajit -b'.

aperezdc committed:
Cater for differences between nDPI 1.7 and 1.8

The API/ABI differences are taken into account so ljndpi works with both
versions of the nDPI library:

* The definition of lib_version is moved to lndpi.c, which now returns a
  table contaning members "version" and "lib" when required. This way
  ljndpi.c can now do different calls to ffi.cdef() depending on the
  version of the loaded nDPI library.

* If an unsupported version of the library is loaded: For pre-1.7 versions,
  an error() is raised, and for post-1.8 versions a warning is issued to
  io.stderr. In the latter case everything might work, but it is not
  guaranteed and warning users seems appropriate.

* The ljndpi.wrap module defines alternative versions of a few functions
  for nDPI 1.8, but the API exposed to Lua remain the same regardless of
  the version of the library loaded.

aperezdc committed:
Import LuaRocks' .rockspec files

These make it possible to install ljnpi using LuaRocks

aperezdc committed:
Release version 0.0.3

aperezdc committed:
Merge commit '4a909ac49f139fe3f850b8f253c68096b39bc161' into snabbwall [ljndpi]

Commit:
Revert "OSX: Switch to Clang as the default compiler."

It breaks cross-compilation to Android.
And host "gcc" aliases to "clang", anyway.

Commit:
Merge branch 'master' into v2.1

aperezdc committed:
Merge tag 'v2016.08' into snabbwall

aperezdc committed:
SnabWall: Add end spy-pcaps test runner

This adds a selftest.sh script for SnabbWall, with one first test suite which
exercises the "snabb wall spy" command using capture files as inputs. For each
capture file, a corresponding .spy.expect file must contain a subset of the L7
protocols expected to be detected by "snabb wall spy", one per line. If any of
the expected L7 protocols is not detected from the capture, the test case is
considered to be failed.

One first test case with a capture file from https://wiki.wireshark.org/SampleCaptures
is provided as well.

aperezdc committed:
SnabbWall: Import a bunch of test cases for "snabb wall spy"

aperezdc committed:
SnabbWall: Print out stats after running "snabb wall spy" tests

aperezdc committed:
Allow iterating over the active flows of a L7 scanner

This adds a new :flows() method to the base apps.wall.scanner class, which
must be implemented in subclasses. It must return an iterator over the active
flows detected by the scanner, and it be suitable to be used in a for-loop:

  for flow in my_scanner:flows() do
     -- do something with "flow"
  end

aperezdc committed:
wall-spy: Report detected flows at the end of non-verbose runs

Previously, not using --verbose/-v would not produce any output at all. This
makes "snabb wall spy" print a summary of the flows at the end of the run when
the flag is not supplied.

aperezdc committed:
wall-spy: Rename -v/--verbose flag to -l/--live

What the flag is actually doing is triggering reporting of flows as they are
detected, so "verbose" is a really bad name for the option. The new name
reflects its purpose much better.

aperezdc committed:
wall-spy: Rename reporter to LiveReporter

LiveReporter better insicates what the application is for.

aperezdc committed:
wall-spy: Document that the command works with TAP and Intel devices

The README for the SnabbWall's "wall spy" command did not reflect the
fact that the tool supports pulling packets from TAP and Intel devices.

aperezdc committed:
wall-spy: Support packet+byte count statistics

In live mode, a line of statistics is produced every two seconds using a
timer, in batch mode a single line summary is printed at the end.

aperezdc committed:
Do not pass "-v" to "snabb wall spy" when running tests

The "-v" command line parameter does not exist anymore, so it has to be
removed. The flows found are now always printed at the end of each batch
run, so no other parameters need to be used. This makes the test suite
work again.

aperezdc committed:
wall-spy: Use correct link name for the RawSocket source

This makes "snabb wall spy raw <device>" actually get packets from the device.

aperezdc committed:
wall-spy: Properly calculate packets/s and bytes/s stats values

The timestamps are obtained with core.app.now(), which already returns values
in seconds. The existing confusion was caused by the fact that timers are
configured using microseconds time values.

aperezdc committed:
Do not leak packets in apps.wall.util.SouthAndNorth

The SouthAndNorth can be used as a kitchen-sink if no outputs are connected.
This means that is must handle deallocating packets that it is not pushing
to the next application in the chain. The changes in this commit also handle
gracefully the case in which one of the :on_{north,south}bound_packet(p) may
return a different packet: the "p" packet is freed and the one returned by the
callback sent instead.

aperezdc committed:
First pass at the SnabbWall application API reference documentation

aperezdc committed:
wall-spy: Only print master protocol when it is known

This makes the output "snabb wall spy" produce less "UNKNOWN" clutter.

aperezdc committed:
wall-spy: Convert port numbers to host endianness for flow reporting

The port numbers where showing oddily because of the missing network-to-host
endianness conversion.

leenaars committed:
Rename snabb to snabbwall

Commit:
Windows/x86: Add MSVC flags for debug build with exception interop.

Commit:
Initialize uv->immutable for upvalues of loaded chunks.

Thanks to Peter Cawley.

Commit:
Merge branch 'master' into v2.1

Commit:
Looks like COLORTERM has gone out of fashion.

Commit:
Merge branch 'master' into v2.1

Commit:
ARM: Fix BLX encoding for Thumb interworking calls.

Thanks to Charles Baylis.

Commit:
Merge branch 'master' into v2.1

Commit:
Fix -jp=a mode for builtins.

Commit:
LJ_GC64: Fix jit.on/off.

Commit:
LJ_GC64: Various fixes.

Contributed by Peter Cawley.

Commit:
Fix GC step size calculation.

Thanks to Igor Ehrlich.

Commit:
Merge branch 'master' into v2.1

Commit:
LJ_FR2: Fix slot 1 handling.

Contributed by Peter Cawley.

aperezdc committed:
Merge tag 'v2016.10' into snabbwall

Commit:
LJ_GC64: Fix HREF for pointers.

Contributed by Peter Cawley.

aperezdc committed:
Use snabb-wall when installed, to reuse Snabb's command resolution logic

When invoked with an argv[0] of snabb-<something>, the snabb binary will
search for a subcommand called <something> and execute it directly.
Then "snabb-wall ..." is equivalent to using "snabb wall ...".

Commit:
LJ_GC64: Various followup fixes.

Contributed by Peter Cawley.

Commit:
Report parent of stitched trace.

Thanks to Nick Zavaritsky.

Commit:
MIPS: Fix TSETR barrier.

Thanks to tongwell.

Commit:
Increase range of GG_State loads via IR_FLOAD with REF_NIL.

Require 32 bit alignment and store offset/4 instead.
Otherwise this can overflow the 10 bit limit for the FOLD op2 key.

Commit:
Fix amalgamated build.

Commit:
Whitespace.

Commit:
ARM64: Add JIT compiler backend.

Contributed by Djordje Kovacevic and Stefan Pejic from RT-RK.com.
Sponsored by Cisco Systems, Inc.

Commit:
ARM64: Allow full VA range for mcode allocation.

Commit:
ARM64: Reject special case in emit_isk13().

Commit:
Generalize deferred constant handling in backend to 64 bit.

Commit:
Update contact info.

Commit:
Merge branch 'master' into v2.1

Commit:
ARM64: Emit more efficient trace exits.

Contributed by Djordje Kovacevic and Stefan Pejic from RT-RK.com.

Commit:
Eliminate use of lightuserdata derived from static data pointers.

Required for >47 bit VA, e.g. ARM64.

Commit:
Document 47 bit limit for lightuserdata.

Commit:
ARM64: Make use of tbz/tbnz and cbz/cbnz.

Contributed by Djordje Kovacevic and Stefan Pejic from RT-RK.com.

takikawa committed:
Merge commit '25289c5797b582a362ae22dad3e1089cc71a0837' into lib/pflua

takikawa committed:
Update Makefile to build dynasm files in pflua

takikawa committed:
Read a "native" config option in pcap_filter

This enables the dynasm backend of pflua when true

takikawa committed:
Extend pcap_filter selftest to test native mode

Commit:
ARM64: Fix pc-relative loads of consts. Cleanup branch codegen.

Thanks to Zhongwei Yao.

Commit:
ARM64: Add missing ldrb/strb instructions to disassembler.

Contributed by Djordje Kovacevic and Stefan Pejic from RT-RK.com.

Commit:
ARM64: Reorganize operand extension definitions.

Commit:
ARM64: Fuse XLOAD/XSTORE with STRREF/ADD/BSHL/CONV.

Commit:
ARM64: Fuse FP multiply-add/sub.

Contributed by Djordje Kovacevic and Stefan Pejic from RT-RK.com.

Commit:
ARM64: Fuse various BAND/BSHL/BSHR/BSAR combinations.

Contributed by Djordje Kovacevic and Stefan Pejic from RT-RK.com.

Commit:
ARM64: Fix code generation for S19 offsets.

Contributed by Zhongwei Yao.

Commit:
Add missing FOLD rule for 64 bit shift+BAND simplification.

Commit:
ARM64: Fuse BOR(BSHL, BSHR) into EXTR/ROR.

Contributed by Djordje Kovacevic and Stefan Pejic from RT-RK.com.

Commit:
ARM64: Use the correct FUSE check.

Oops, my bad.

Commit:
Add "proto" field to jit.util.funcinfo().

Commit:
Add "proto" field to jit.util.funcinfo().

Backport.

Commit:
Merge branch 'master' into v2.1

Commit:
ARM64: Fuse BOR/BXOR and BNOT into ORN/EON.

Contributed by Djordje Kovacevic and Stefan Pejic from RT-RK.com.

Commit:
ARM64: Remove unused variables in disassembler.

Thanks to François Perrad.

Commit:
Merge PR #1088 (Update pflua) into kbara-next

Commit:
Fix cross-endian jit.bcsave for MIPS target.

Commit:
Merge branch 'master' into v2.1

Commit:
Fix HTML formatting.

Commit:
Merge branch 'master' into v2.1

Commit:
Add some more extensions from Lua 5.2/5.3.

Contributed by François Perrad.

Commit:
LJ_GC64: Add build options and install instructions.

Commit:
Bump copyright date to 2017.

Commit:
Merge branch 'master' into v2.1

Commit:
Fix extension docs about package.searchers.

aperezdc committed:
Docs: Describe Scanner:extract_packet_info() and key objects

This adds some more details in the description of Scanner objects, and
documents the :extract_packet_info() method, including the "key objects"
returned by it.

takikawa committed:
Merge tag 'v2016.11' into snabbwall

Conflicts:
	src/doc/genbook.sh

takikawa committed:
Fix references to intel1/10g apps in `wall spy`

takikawa committed:
Merge pull request #13 from aperezdc/apidocs

Docs: Describe Scanner:extract_packet_info() and key objects

takikawa committed:
First attempt at a l7fw app and filter program

takikawa committed:
Allow pfmatch actions for l7fw rules

takikawa committed:
Modify pflua to support extra pfmatch arguments

Also undo the handler caching code from the last few commits.
It looks like modifying pflua is a better solution, even if
it's hacky, than trying to compile on the fly (it's too slow).

takikawa committed:
Backport pflua fixes from Igalia/pflua

These fixes are already in upstream pflua but not
yet in a snabb release

takikawa committed:
Allow pfmatch rules to use the flow packet count

This compiles a pfmatch expression for each flow that
matches a pfmatch rule in the firewall rules. Since
there are usually several packets in a flow, this amortizes
the compilation cost across the packets.

This is ok if flows typically have many packets, but
that's not always the case. Unfortunately, it's not
straightforward to compile them ahead of time when using
the $flow_count variable without invasive changes to
pflua.

It's possible to change it so that it only compiles on
each flow if the $flow_count variable is used.

takikawa committed:
Have to index handler table by name and count

It's not enough to index by the flow identity since
its count is mutated (and we need to recompile for
different counts).

takikawa committed:
Only recompile if the pfmatch uses $flow_count

takikawa committed:
Refactor spy program to allow code sharing

takikawa committed:
Make snabb wall filter more complete

Adds multiple input sources, flags to set the output
path and whether to show the link report, and reads
the firewall rules from a file.

takikawa committed:
First implementation of reject l7fw policy

takikawa committed:
Send ICMP responses only if host ip/mac are set

takikawa committed:
Output a TCP RST for rejects for TCP traffic

takikawa committed:
Fix typo in lib.protocol.tcp constructor

Submitted to upstream snabb separately

takikawa committed:
Add README for `snabb wall filter`

takikawa committed:
Send rejection packets to a separate output port

takikawa committed:
Add support for ICMPv6 responses in l7fw

takikawa committed:
Enable passing IP & MAC to `snabb wall filter`

takikawa committed:
Make `snabb config filter` take a rules argument

Either -e or -f specifies the firewall rules. -e for a
string that specifies the rules, -f for a file.

takikawa committed:
Add a test script for 'snabb wall filter'

takikawa committed:
Fix undefined variable in pf.savefile

Merged in upstream pflua as d6a0f167881800871f736745837bece10edebd14

takikawa committed:
Fix incorrect ethernet type for IPv6

takikawa committed:
Fix incorrect payload calculation for ICMPv6

takikawa committed:
Add unit tests for apps.wall.l7fw

takikawa committed:
Add docs for L7Spy and L7Fw

takikawa committed:
Adjust -e and -f flags and document both

takikawa committed:
Add a report option to `snabb config filter`

Replaces the former -l option that showed the links and
instead prints the accepeted, rejected, dropped counts
from the l7fw app.

takikawa committed:
Update test for new `snabb wall filter` report

takikawa committed:
Support logging packet actions to the system log

Currently there are only "on" and "off" log levels.
The "on" log level, like ufw's "on" or "low" modes,
only logs DROP/REJECT actions.

takikawa committed:
Simplify handling of extra pflua filter arguments

Instead of using an `extra_args` table with separate
logic for tracking the `len` variable, keep the `len`
variable in the table too.

Also localize `filter_args` in each module

takikawa committed:
Move assert statement earlier in the call graph

takikawa committed:
Lift some constants as definitions

Also revise/add a few relevant comments

takikawa committed:
Assert that the rules file exists

takikawa committed:
Explicitly require the bit library in l7fw

takikawa committed:
Fix typo in L7Spy API docs

takikawa committed:
Merge pull request #14 from takikawa/snabbwall-dev

Add l7fw app and `snabb wall filter`

takikawa committed:
Workaround for intermittent segfaults

For unknown reasons, the `newp == false` check triggers
a segfault when running Snabbwall in certain cases. This
change to the conditional guard is a workaround until the
deeper issue is fixed.

takikawa committed:
Add duration arguments for 'snabb wall` commands

takikawa committed:
Fix "tx"/"rx" link names that were reversed

takikawa committed:
Fix PPS & B/s printouts for `snabb wall spy`

When only -s is passed, the PPS and B/s fields in
the report were wrong because the start time was set to 0
rather than the current time.

takikawa committed:
Move constants to separate module & localize

Improves performance of `snabb wall spy` by a good amount

takikawa committed:
Pre-allocate the flow key objects and rewrite

This improves performance quite a bit by eliminating an
allocation that causes a trace abort. The caveat is that
Snabbwall must ensure that the lifetime of the flow keys
are short, so that conflicting writes do not happen.

takikawa committed:
Use dynasm to optimize ndpi binding functions

Dynasm is used to construct a wrapper function that avoids a
struct return (that LuaJIT does not support for traced FFI
calls) by using a struct pointer.

We use dynasm to avoid having to compile a C wrapper which
complicates the build process. The wrapper isn't expressible
in Lua.

takikawa committed:
Add a --cpu option to 'snabb wall filter'

takikawa committed:
Report bytes/packets/bps for 'snabb config filter'

Commit:
ARM64: Fix AREF/HREF/UREF fusion.

Thanks to Zhongwei Yao.

takikawa committed:
Add a benchmarking script for 'snabb wall filter'

Commit:
Make checkptrGC() actually work.

Neither LJ_64 nor LJ_GC64 are defined when lj_def.h is included.
So we'll need to use lazy C macro evaluation.

Commit:
MIPS: Backport workaround for compact unwind tables.

Commit:
MIPS: Fix emitted code for U32 to float conversion.

Commit:
MIPS: Don't use RID_GP as a scratch register.

Commit:
Merge branch 'master' into v2.1

Commit:
ARM64: Cleanup and de-cargo-cult TValue store generation.

Commit:
x64/LJ_GC64: Fix (currently unused) integer stores in asm_tvptr().

Commit:
x64/LJ_GC64: Fix warning for DUALNUM build.

Commit:
Fix soft-float math.abs() and negation.

Broken since SIMD constants were switched to IR_FLOAD REF_NIL.

Commit:
Fix FOLD rules for math.abs() and FP negation.

Broken since SIMD constants were switched to IR_FLOAD REF_NIL.

Commit:
MIPS64, part 2: Add MIPS64 hard-float JIT compiler backend.

Contributed by Djordje Kovacevic and Stefan Pejic from RT-RK.com.
Sponsored by Cisco Systems, Inc.

takikawa committed:
Fix typo ("nddp" -> "ndpp")

takikawa committed:
Merge pull request #15 from takikawa/snabbwall-dev

Optimizations for snabbwall & benchmarking script

takikawa committed:
Adjust snabbwall test path to work on NixOS

takikawa committed:
Change binary from "snabbwall" -> "snabb-wall"

Closes Snabbwall issue #16

takikawa committed:
Update .gitignore for ljndpi

takikawa committed:
Revert Makefile changes for upstreaming Snabbwall

Undoes some Makefile changes that were specific to Snabbwall
development that don't make sense for upstream

lukego committed:
Merge v2017.03 branch 'snabbco/master' into next

Commit:
DynASM: Fix warning.

Commit:
Fix annoying warning, due to deterministic binutils configuration.

Commit:
Remove internal __mode = "K" and replace with safe check.

Commit:
Fix out-of-scope goto handling in parser.

Many thanks to Demetrios Obenour for tracking down this long-standing bug.

Commit:
Fix overly restrictive range calculation in mcode allocation.

Contributed by Alexey Kopytov.

Commit:
Limit mcode alloc probing, depending on the available pool size.

Contributed by Alexey Kopytov.

Commit:
Remove unnecessary mcode alloc pointer check.

Also fixes Illumos address space issue reported by Theo Schlossnagle.

Commit:
Merge branch 'master' into v2.1

lukego committed:
Squashed 'lib/luajit/' changes from 2f6b296..d3e36e7

d3e36e7 Merge branch 'master' into v2.1
f50bf75 Remove unnecessary mcode alloc pointer check.
d62459f Limit mcode alloc probing, depending on the available pool size.
4e30836 Fix overly restrictive range calculation in mcode allocation.
ff64836 Fix out-of-scope goto handling in parser.
3ab9f5a Remove internal __mode = "K" and replace with safe check.
eef77a6 Fix annoying warning, due to deterministic binutils configuration.
a9740d9 DynASM: Fix warning.
a25c0b9 MIPS64, part 2: Add MIPS64 hard-float JIT compiler backend.
4416e88 Fix FOLD rules for math.abs() and FP negation.
019fb9d Fix soft-float math.abs() and negation.
130d1dc x64/LJ_GC64: Fix warning for DUALNUM build.
f640ec7 x64/LJ_GC64: Fix (currently unused) integer stores in asm_tvptr().
0a46ef1 ARM64: Cleanup and de-cargo-cult TValue store generation.
d0759e4 Merge branch 'master' into v2.1
892d370 MIPS: Don't use RID_GP as a scratch register.
ee33a1f MIPS: Fix emitted code for U32 to float conversion.
1abd779 MIPS: Backport workaround for compact unwind tables.
6bf3e4d Make checkptrGC() actually work.
5aa0201 ARM64: Fix AREF/HREF/UREF fusion.
bd7e42e Fix extension docs about package.searchers.
71ff7ef Merge branch 'master' into v2.1
b93a1dd Bump copyright date to 2017.
c94b921 LJ_GC64: Add build options and install instructions.
c198167 Add some more extensions from Lua 5.2/5.3.
ed4ce98 Merge branch 'master' into v2.1
a1e13fa Fix HTML formatting.
12c0df4 Merge branch 'master' into v2.1
a2013dd Fix cross-endian jit.bcsave for MIPS target.
8e5d7be ARM64: Remove unused variables in disassembler.
ebec253 ARM64: Fuse BOR/BXOR and BNOT into ORN/EON.
3cfa9cb Merge branch 'master' into v2.1
fb61f7c Add "proto" field to jit.util.funcinfo().
1973807 Add "proto" field to jit.util.funcinfo().
4ccd876 ARM64: Use the correct FUSE check.
44b99ff ARM64: Fuse BOR(BSHL, BSHR) into EXTR/ROR.
ec2756b Add missing FOLD rule for 64 bit shift+BAND simplification.
986854c ARM64: Fix code generation for S19 offsets.
3975b6c ARM64: Fuse various BAND/BSHL/BSHR/BSAR combinations.
2772cbc ARM64: Fuse FP multiply-add/sub.
bfeb116 ARM64: Fuse XLOAD/XSTORE with STRREF/ADD/BSHL/CONV.
2ac2cd4 ARM64: Reorganize operand extension definitions.
48b0029 ARM64: Add missing ldrb/strb instructions to disassembler.
22511fb ARM64: Fix pc-relative loads of consts. Cleanup branch codegen.
3ad2bbf ARM64: Make use of tbz/tbnz and cbz/cbnz.
6538c8a Document 47 bit limit for lightuserdata.
d7243e1 Eliminate use of lightuserdata derived from static data pointers.
8125989 ARM64: Emit more efficient trace exits.
1131fa2 Merge branch 'master' into v2.1
c3cae04 Update contact info.
a566544 Generalize deferred constant handling in backend to 64 bit.
2b77da3 ARM64: Reject special case in emit_isk13().
7a0c3a1 ARM64: Allow full VA range for mcode allocation.
04b6070 ARM64: Add JIT compiler backend.
13642b7 Whitespace.
202713a Fix amalgamated build.
e577db5 Increase range of GG_State loads via IR_FLOAD with REF_NIL.
5400c1e MIPS: Fix TSETR barrier.
7a58a8f Report parent of stitched trace.
716f2da LJ_GC64: Various followup fixes.
bdcaf4b LJ_GC64: Fix HREF for pointers.
6a25014 LJ_FR2: Fix slot 1 handling.
3f43f09 Merge branch 'master' into v2.1
a68c411 Fix GC step size calculation.
54b78e7 LJ_GC64: Various fixes.
63465fe LJ_GC64: Fix jit.on/off.
cf80edb Fix -jp=a mode for builtins.
f27b250 Merge branch 'master' into v2.1
fcc8244 ARM: Fix BLX encoding for Thumb interworking calls.
4ca3909 Merge branch 'master' into v2.1
8ada57e Looks like COLORTERM has gone out of fashion.
1a9a264 Merge branch 'master' into v2.1
9910ded Initialize uv->immutable for upvalues of loaded chunks.
03b03ef Windows/x86: Add MSVC flags for debug build with exception interop.
f6f838f Merge branch 'master' into v2.1
02b9b55 Revert "OSX: Switch to Clang as the default compiler."
972a1a4 Fix exit status for 'luajit -b'.
c98660c Must preserve J->fold.ins (fins) around call to lj_ir_ksimd().
22e8e07 Merge branch 'master' into v2.1
d41469c Emit bytecode in .c/.h files with unsigned char type.
92d9ff2 Set arg table before evaluating LUA_INIT and -e chunks.
6be5ffd Adjust comment with defines.
7374046 Merge branch 'master' into v2.1
b74ddaf Fix for cdata vs. non-cdata arithmetics/comparisons.
37b377d Merge branch 'master' into v2.1
1914de7 Fix unused vars etc. in internal Lua files.
01e4754 Properly clean up state before restart of trace assembly.
ce30766 Merge branch 'master' into v2.1
aef4edd Drop leftover regs in 'for' iterator assignment, too.
287a534 MIPS: Support MIPS16 interlinking.
f598343 x64/LJ_GC64: Fix code generation for IR_KNULL call argument.
a88dc57 Merge branch 'master' into v2.1
a7bec69 Fix PHI remarking in SINK pass.
ce1ad87 LJ_GC64: Set correct nil value when clearing a cdata finalizer.
cc05e79 LJ_GC64: Ensure all IR slot fields are initialized.
58ca165 LJ_GC64: Allow optional use of the system memory allocator.
7d43402 Merge branch 'master' into v2.1
3206307 Fix Valgrind suppressions.
1c0454c Merge branch 'master' into v2.1
a5f8a48 Don't try to record outermost pcall() return to lower frame.
384ce2f MIPS: Fix build failures and warnings.
56fe899 Proper fix for LJ_GC64 changes to asm_href().
d9986fb MIPS64, part 1: Add MIPS64 support to interpreter.
e3c4c9a DynASM/MIPS: Add missing MIPS64 instructions.
5e2b609 Fix compiler warning.
e77638f x64/LJ_GC64: Fix __call metamethod for tailcall.
6360f6e Fix collateral damage from LJ_GC64 changes to asm_href().
9fa843a Cleanup install docs.
3f1031c Use MAP_TRYFIXED for the probing memory allocator, if available.
bfe2a35 Merge branch 'master' into v2.1
747feb6 x86: Don't spill an explicit REF_BASE in the IR.
2868715 x64/LJ_GC64: Add missing backend support and enable JIT compilation.
6c8258d LJ_FR2: Add support for trace recording and snapshots.
8f868a9 LJ_GC64: Update IR type sizes.
f26679c LJ_GC64: Add support for 64 bit GCobj constants in the IR.
9e99ccc Strip out old infrastructure for 64 bit constants.
7fb75cc Embed 64 bit constants directly in the IR, using two slots.
a406797 Always walk IR constants in ascending order.
3152ed9 Simplify GCtrace * reference embedding for trace stitching.
a657fa0 Make the IR immovable after assembly.
5135876 Add ra_addrename().
ccae333 Load SIMD constants with IR_FLOAD from GG_State.
786dbb2 Add IR_FLOAD with REF_NIL for field loads from GG_State.
cfa188f Move common 32/64 bit in-memory FP constants to jit_State.
1931b38 LJ_GC64: Introduce IRT_PGC.
475a6ae Merge branch 'master' into v2.1
37e1e70 Add guard for obscure aliasing between open upvalues and SSA slots.
d4f3b11 Workaround for MinGW headers lacking some exception definitions.
4fe400c Merge branch 'master' into v2.1
5837c2a Remove assumption that lj_math_random_step() doesn't clobber FPRs.
573daa9 Fix dependencies.
35b09e6 Windows/x86: Add full exception interoperability.
6a99732 Merge branch 'master' into v2.1
f05280e x86/x64: Fix instruction length decoder.
221268b Use the GDB JIT API in a thread-safe manner.
ac42037 Constrain value range of lj_ir_kptr() to unsigned 32 bit pointers.
d8ac623 Merge branch 'master' into v2.1
7b26e9c Fix GCC 6 -Wmisleading-indentation warnings.
344fe5f Merge branch 'master' into v2.1
2f0001f Fix handling of non-numeric strings in arithmetic coercions.
4c6498d Merge branch 'master' into v2.1
cc4f5d0 Whitespace.
d13d420 Merge branch 'master' into v2.1
73680a5 x86/x64: Search for exit jumps with instruction length decoder.
0c6fdc1 Rewrite memory block allocator.
101115d Merge branch 'master' into v2.1
e5b5e07 MIPS: Fix BC_ISNEXT fallback path.
096a7cf x64/LJ_GC64: Fix BC_UCLO check for fast-path.
ac9193c x86: Improve disassembly of BMI2 instructions.
d150fbf Merge branch 'master' into v2.1
1c6fd13 Fix recording of select(n, ...) with off-trace varargs
25b3779 Merge branch 'master' into v2.1
4ab6367 Cygwin: Allow cross-builds to non-Cygwin targets.
296f0ca Windows/x64/LJ_GC64: Fix math.frexp() and math.modf() (again).
6e623b9 Merge branch 'master' into v2.1
62af101 MIPS: Fix use of ffgccheck delay slots in interpreter.
892887e x86: Generate BMI2 shifts and rotates, if available.
6801e71 x86: Detect BMI2 instruction support.
c24c8e5 x64/LJ_GC64: Fix JIT glue code in interpreter.
d714561 Merge branch 'master' into v2.1
9531eb2 Windows: Remove intermediate files at end of build.
e03e597 Fix compiler warnings.
df7bb5b Merge branch 'master' into v2.1
e23fc10 Fix display of NULL (light)userdata in -jdump.
c730540 Fix formatting of some small denormals at low precision.
713e340 Merge branch 'master' into v2.1
7e05355 Fix install for cross-builds.
cf3e01e Merge branch 'master' into v2.1
ddadbe8 Fix compiler warnings.
be25592 ARM: Fix build problem with external frame unwinding.
64c6da6 MIPS soft-float: Fix code generation for HREF.
75d046d DynASM/x86: Add ADX instructions.
3e4a196 RELEASE LuaJIT-2.1.0-beta2
8290076 Update changelog.
f423194 Merge branch 'master' into v2.1
db1b399 Bump copyright date to 2016.
6f3eed9 Fix Android/x86 build.
1c45c6a Fix build with JIT disabled.
18f6aa9 Use internal implementation for converting FP numbers to strings.
339a1fd Windows/x64/LJ_GC64: Fix math.frexp() and math.modf().
6cb38f7 Merge branch 'master' into v2.1
a443889 Don't allocate unused 2nd result register in JIT compiler backend.
f547a14 MIPS: Add soft-float support to JIT compiler backend.
825dcdc Generalize LJ_SOFTFP dependencies in lj_asm.c.
bfbcd72 PS4: Switch default build to amalgamated and LJ_GC64 mode.
60de2f3 MIPS: Switch to dual-number mode. Fix soft-float interpreter.

git-subtree-dir: lib/luajit
git-subtree-split: d3e36e7920c641410dfcdf1fc6c10069fd3192a6

lukego committed:
Merge commit '736cc366405a366cae3760e1b1a9e6e7ea3871db' into luajit-upgrade

lukego committed:
Merge commit 'd3e36e7920c641410dfcdf1fc6c10069fd3192a6' into luajit-upgrade

This commit affects only the git history and not the contents of the code.

The unsquashed LuaJIT commit history is merged into our lib/luajit
subtree. This should make it easier for us to pull in changes to
LuaJIT from multiple branches, because git subtree will know which
commits we have and which ones we don't.

The code is left unchanged. There were some conflicts where we had
changed something compared with LuaJIT upstream but I resolved all of
those in favor of keeping our code the same. (If we actually don't
want changes then we can push reverts later.)

lukego committed:
Makefile: Bump luajit version number to beta2

lukego committed:
doc/branches.md: Add 'luajit' branch

lukego committed:
Merge #1111 branch 'lukego/luajit-upgrade-history' into luajit

lukego committed:
Merge #1113 branch 'lukego/luajit-branch' into next

lukego committed:
Merge #1111 (via branch 'snabbco/luajit') into next

lukego committed:
Merge #1090 branch 'snabbco/kbara-next' into next

takikawa committed:
Merge tag 'v2017.03' into snabbwall

takikawa committed:
Add an ok printout for l7fw & scanner selftests

kbara commented:
Hi Asumu.

Overall, this looks great. Could you open a new PR against snabb/next rather than kbara/kbara-next? It's a bit more convenient with the tooling. :-)

I'm happy to have the pflua changes be part of the same PR; they're both code that Igalia effectively controls. The changes to non-Igalia-owned files look minimal and necessary.

I'll ping lukego tomorrow about whether ndpi licensing is an issue. If it's not, I'll take another look at this with fresh eyes, but to me it looks mergeable as-is. :-)

takikawa committed:
Add Snabbwall branch info to src/doc/branches.md

takikawa committed:
Merge branch 'upstream/next' into snabbwall

Usually Snabbwall has tracked release versions of Snabb.
I'm merging in the `next` branch instead now in order to
make it easier to upstream Snabbwall without merge
conflicts in the pull request.

Conflicts:
	lib/pflua/src/pf/anf.lua
	lib/pflua/src/pf/match.lua

aperezdc committed:
Relicense to Apache License 2.0

One of the main motivations of ljndpi is to be used as part of
SnabbWall, a suite of Layer 7 firewall components for Snabb, which
itself is Apache-licensed. Using the same license as Snabb makes it
simpler to incorporate as part of it.

aperezdc committed:
Release version 0.1.0

takikawa committed:
Merge commit '9c08985a28777d1818547b4f7c024b7163505687' into snabbwall
